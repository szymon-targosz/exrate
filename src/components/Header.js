import React from 'react'

const Header = () => (
   <header className="header">
      <div className="container">
         <h1 className="header__title">Exrate</h1>
      </div>
   </header>
)

export default Header;
