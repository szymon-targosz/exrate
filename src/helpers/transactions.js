import numeral from 'numeral';
import 'numeral/locales';

const getHighest = transactions =>
   transactions.length === 0 ? undefined : transactions.reduce((prev, curr) => (prev.amount > curr.amount ? prev : curr), []);

const getTotal = transactions => transactions.map(transaction => transaction.amount).reduce((sum, val) => sum + val, 0);

const selectTransactions = (transactions, { text, sort }) => {
   return transactions
      .filter(transaction => {
         const textMatch = transaction.name.toLowerCase().includes(text.toLowerCase());
         return textMatch;
      })
      .sort((a, b) => {
         if (sort === 'alpha') {
            return a.name > b.name ? 1 : -1;
         } else if (sort === 'asc') {
            return a.amount - b.amount;
         } else {
            return b.amount - a.amount;
         }
      });
};

const generateAmountMsg = (amount, locale = 'de', decsFit = false) => {
   numeral.locale(locale);

   let decNums = '00';
   if (decsFit) {
      const decs = amount.toString().split('.')[1] || '';
      let i = 0;
      while (i < decs.length - 2) {
         decNums += 0;
         i++;
      }
   }

   return numeral(amount).format(`0,0.${decNums} $`);
};

export { getHighest, getTotal, selectTransactions, generateAmountMsg };
