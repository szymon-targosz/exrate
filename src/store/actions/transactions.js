import uuidv4 from 'uuid/v4';
import { ADD_TRANSACTION, REMOVE_TRANSACTION, REMOVE_ALL_TRANSACTIONS } from './actionTypes';

const addTransaction = (name, amount) => ({
   type: ADD_TRANSACTION,
   transaction: {
      id: uuidv4(),
      amount: parseFloat(amount),
      name
   }
});

const removeTransaction = id => ({
   type: REMOVE_TRANSACTION,
   id
});

const removeAllTransactions = () => ({ type: REMOVE_ALL_TRANSACTIONS });

export { addTransaction, removeTransaction, removeAllTransactions };
