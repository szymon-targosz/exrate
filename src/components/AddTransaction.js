import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addTransaction } from '../store/actions';
import SectionHeader from './UI/SectionHeader';
import Input from './UI/Input';

export class AddTransaction extends Component {
   state = { name: '', amount: '', error: undefined };
   inputNameRef = React.createRef();

   handleSubmit = e => {
      e.preventDefault();
      const name = this.state.name.trim();
      if (name === '') return this.setState({ error: 'Please enter the transaction name' });

      const isOccupied = this.props.transactions.find(t => t.name === name);
      if (isOccupied) return this.setState({ error: 'Name already taken' });

      this.props.addTransaction(name, this.state.amount);
      this.setState({ error: undefined, name: '', amount: '' });
      this.inputNameRef.current.focus();
   };

   onNameChange = e => {
      this.setState({ name: e.target.value });
   };

   onAmountChange = e => {
      const amount = e.target.value;
      if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
         this.setState({ amount });
      }
   };

   render() {
      return (
         <section className="add-transaction">
            <SectionHeader title="Add transaction" />
            <form onSubmit={this.handleSubmit} className="p-s">
               <Input
                  elementType="input"
                  changeHandler={this.onNameChange}
                  config={{
                     name: 'name',
                     ref: this.inputNameRef,
                     value: this.state.name,
                     placeholder: 'Name',
                     autoComplete: 'off',
                     id: 'name-input',
                     type: 'text'
                  }}
                  label
               />
               <Input
                  elementType="input"
                  changeHandler={this.onAmountChange}
                  config={{
                     name: 'amount',
                     value: this.state.amount,
                     placeholder: 'Amount (EUR)',
                     autoComplete: 'off',
                     id: 'amount-input',
                     type: 'text'
                  }}
                  label
               />
               {this.state.error && <p className="error">{this.state.error}</p>}
               <button className="btn btn--block" disabled={!this.state.name || !this.state.amount}>
                  Add
               </button>
            </form>
         </section>
      );
   }
}

const mapStateToProps = ({ transactions }) => ({ transactions });

const mapDispatchToProps = dispatch => ({
   addTransaction: (name, amount) => dispatch(addTransaction(name, amount))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTransaction);
