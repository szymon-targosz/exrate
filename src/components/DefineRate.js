import React, { Component } from 'react';
import { connect } from 'react-redux';
import SectionHeader from './UI/SectionHeader';
import InfoBox from './UI/InfoBox';
import Input from './UI/Input';
import { setExchangeRate } from '../store/actions';
import { generateAmountMsg } from '../helpers/transactions';

export class DefineRate extends Component {
   state = { rate: 0 };

   handleInput = e => {
      const rate = e.target.value;
      if (!rate || rate.match(/^\d{1,}(\.\d{0,9})?$/)) {
         this.setState({ rate: rate });
         this.props.setExchangeRate(rate || 0);
      }
   };

   render() {
      return (
         <section className="define-rate">
            <SectionHeader title="Set exchange rate" cssClasses="define-rate__header" />
            <div className="define-rate__form p-s">
               <Input
                  elementType="input"
                  cssClasses="input__control--short"
                  changeHandler={this.handleInput}
                  config={{
                     value: this.state.rate,
                     placeholder: 'Exchange rate',
                     id: 'rate-input',
                     type: 'text'
                  }}
                  label
               />
            </div>
            <div className="define-rate__current">
               <InfoBox
                  description="Current exchange rate"
                  data={`${generateAmountMsg(1)} = ${generateAmountMsg(parseFloat(this.state.rate) || 0, 'pl', true)}`}
               />
            </div>
         </section>
      );
   }
}

const mapDispatchToProps = dispatch => ({ setExchangeRate: rate => dispatch(setExchangeRate(rate)) });

export default connect(null, mapDispatchToProps)(DefineRate);
