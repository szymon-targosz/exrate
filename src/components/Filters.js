import React, { Component } from 'react';
import SectionHeader from './UI/SectionHeader';
import Input from './UI/Input';
import { connect } from 'react-redux';
import { setTextFilter, sortAlpha, sortAsc, sortDesc } from '../store/actions';

export class Filters extends Component {
   onTextChange = e => {
      this.props.setTextFilter(e.target.value);
   };

   onSortChange = e => {
      const sort = e.target.value;
      if (sort === 'alpha') {
         this.props.sortAlpha();
      } else if (sort === 'desc') {
         this.props.sortDesc();
      } else {
         this.props.sortAsc();
      }
   };

   render() {
      return (
         <section className="filters">
            <SectionHeader title="Filters" />
            <div className="p-s">
               <Input
                  elementType="input"
                  config={{
                     placeholder: 'Search',
                     value: this.props.filters.text,
                     id: 'search-input',
                     type: 'search',
                     autoComplete: 'off'
                  }}
                  changeHandler={this.onTextChange}
                  label
               />
               <Input
                  elementType="select"
                  cssClasses="input__control--select"
                  config={{ value: this.props.filters.sort }}
                  changeHandler={this.onSortChange}
                  options={[
                     { value: 'alpha', text: 'Alphabetically' },
                     { value: 'asc', text: 'Ascending' },
                     { value: 'desc', text: 'Descending' }
                  ]}
               />
            </div>
         </section>
      );
   }
}

const mapStateToProps = ({ filters }) => ({ filters });

const mapDispatchToProps = dispatch => ({
   setTextFilter: text => dispatch(setTextFilter(text)),
   sortAlpha: () => dispatch(sortAlpha()),
   sortAsc: () => dispatch(sortAsc()),
   sortDesc: () => dispatch(sortDesc())
});

export default connect(mapStateToProps, mapDispatchToProps)(Filters);
