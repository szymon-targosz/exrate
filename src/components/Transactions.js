import React from 'react';
import { connect } from 'react-redux';
import { removeTransaction, removeAllTransactions } from '../store/actions';
import Transaction from './Transaction';
import TransactionsTotal from './TransactionsTotal';
import { getTotal } from '../helpers/transactions';
import SectionHeader from './UI/SectionHeader';
import { selectTransactions } from '../helpers/transactions';

export const Transactions = ({ removeTransaction, removeAllTransactions, transactions, rate, total }) => (
   <section className="transactions">
      <SectionHeader cssClasses="transactions__header" title="Transactions">
         <button className="btn btn--link" onClick={removeAllTransactions} disabled={!(transactions.length > 0)}>
            Remove All
         </button>
      </SectionHeader>
      <div className="transactions__content">
         <div className="transactions__list">
            {transactions.map(transaction => (
               <Transaction
                  key={transaction.id}
                  removeHandler={() => removeTransaction(transaction.id)}
                  transaction={transaction}
                  rate={rate}
               />
            ))}
            {transactions.length === 0 && <p className="transactions__info p-m">There is no transaction</p>}
         </div>
         <TransactionsTotal total={total} rate={rate} quantity={transactions.length} />
      </div>
   </section>
);

const mapStateToProps = ({ transactions, rate, filters }) => {
   const selected = selectTransactions(transactions, filters);
   const total = getTotal(selected);
   return { transactions: selected, rate, total };
};

const mapDispatchToProps = dispatch => ({
   removeTransaction: id => dispatch(removeTransaction(id)),
   removeAllTransactions: () => dispatch(removeAllTransactions())
});

export default connect(mapStateToProps, mapDispatchToProps)(Transactions);
