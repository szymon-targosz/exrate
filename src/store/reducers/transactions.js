import { ADD_TRANSACTION, REMOVE_TRANSACTION, REMOVE_ALL_TRANSACTIONS } from '../actions/actionTypes';

export default (state = [], action) => {
   switch (action.type) {
      case ADD_TRANSACTION:
         return state.concat(action.transaction);

      case REMOVE_TRANSACTION:
         return state.filter(({ id }) => id !== action.id);

      case REMOVE_ALL_TRANSACTIONS:
         return [];

      default:
         return state;
   }
};
