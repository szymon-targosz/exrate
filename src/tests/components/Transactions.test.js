import React from 'react';
import { Transactions } from '../../components/Transactions';
import { shallow } from 'enzyme';
import transactions from '../fixtures/transactions';

let wrapper, removeTransaction, removeAllTransactions;

beforeEach(() => {
   removeTransaction = jest.fn();
   removeAllTransactions = jest.fn();
   wrapper = shallow(
      <Transactions
         transactions={transactions}
         removeTransaction={removeTransaction}
         removeAllTransactions={removeAllTransactions}
         rate={4.46}
         total={451.69}
      />
   );
});

describe('COMPONENT - Transactions', () => {
   test('Should render Transactions correctly', () => {
      expect(wrapper).toMatchSnapshot();
   });

   test('Should render Transactions with empty message', () => {
      wrapper.setProps({ transactions: [], rate: 0, total: 0 });
      expect(wrapper).toMatchSnapshot();
   });

   test('Should handle removeAllTransactions', () => {
      wrapper.find('button').simulate('click');
      expect(removeAllTransactions).toHaveBeenCalled();
   });
});
