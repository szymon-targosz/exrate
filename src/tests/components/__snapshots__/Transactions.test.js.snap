// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - Transactions Should render Transactions correctly 1`] = `
<section
  className="transactions"
>
  <SectionHeader
    cssClasses="transactions__header"
    title="Transactions"
  >
    <button
      className="btn btn--link"
      disabled={false}
      onClick={[MockFunction]}
    >
      Remove All
    </button>
  </SectionHeader>
  <div
    className="transactions__content"
  >
    <div
      className="transactions__list"
    >
      <Transaction
        key="1"
        rate={4.46}
        removeHandler={[Function]}
        transaction={
          Object {
            "amount": 41,
            "id": "1",
            "name": "Bills",
          }
        }
      />
      <Transaction
        key="2"
        rate={4.46}
        removeHandler={[Function]}
        transaction={
          Object {
            "amount": 299.99,
            "id": "2",
            "name": "new phone",
          }
        }
      />
      <Transaction
        key="3"
        rate={4.46}
        removeHandler={[Function]}
        transaction={
          Object {
            "amount": 110.7,
            "id": "3",
            "name": "headphones",
          }
        }
      />
    </div>
    <TransactionsTotal
      quantity={3}
      rate={4.46}
      total={451.69}
    />
  </div>
</section>
`;

exports[`COMPONENT - Transactions Should render Transactions with empty message 1`] = `
<section
  className="transactions"
>
  <SectionHeader
    cssClasses="transactions__header"
    title="Transactions"
  >
    <button
      className="btn btn--link"
      disabled={true}
      onClick={[MockFunction]}
    >
      Remove All
    </button>
  </SectionHeader>
  <div
    className="transactions__content"
  >
    <div
      className="transactions__list"
    >
      <p
        className="transactions__info p-m"
      >
        There is no transaction
      </p>
    </div>
    <TransactionsTotal
      quantity={0}
      rate={0}
      total={0}
    />
  </div>
</section>
`;
